### PR TITLE
Configuration based TUDelft calculation

### PR DESCRIPTION
**Pull request by joemoorhouse**
Finish changes related to running a flood calculation using TUDelft data, including standard of protection, and using configuration-based vulnerability.
Partial implementation of vulnerability function uncertainty in config-based models and add to docs.

### CODE REVIEW

**Review by xbarra — CHANGES_REQUESTED** [review]
Please, rebase this branch onto main before merging.

**Review by xbarra — CHANGES_REQUESTED** [review]
I would rather this got moved to it's own pr and merged.

Also a part of a bigger conversation, but squashing it on this merge would make it difficult to track

File: src/physrisk/vulnerability_models/config_based_impact_curves.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+                        }
+                    )
+                )
+                if reduced_key in keys:
+                    return reduced_key
+            raise ValueError("Failed to map asset to an impact curve.")
+        return key
+
+
+class ImpactCurve:
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_impact_curves.py:218]: Missing docstring in public class(fails Ruff rule D101)

File: src/physrisk/vulnerability_models/config_based_impact_curves.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    def __init__(
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_impact_curves.py:219]: Missing docstring in __init__ (fails Ruff rule D107)

File: src/physrisk/vulnerability_models/config_based_impact_curves.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+    ):
+        assert curve_type in ["indicator", "threshold"]
+        self.curve_type = curve_type
+        self.indicator_units = indicator_units
+        self.cap_of_points_x = cap_of_points_x
+        self.cap_of_points_y = cap_of_points_y
+        self.activation_of_points_x = activation_of_points_x
+
+    @abstractmethod
+    def get_impact(self, points_x: np.ndarray) -> np.ndarray: ...
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_impact_curves.py:235]: Missing docstring in public method (fails Ruff rule D102)

File: src/physrisk/vulnerability_models/config_based_impact_curves.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+    @staticmethod
+    def get_impact_curve(item: VulnerabilityConfigItem):
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_impact_curves.py:238]: Missing docstring in public method (fails Ruff rule D102)

File: src/physrisk/vulnerability_models/config_based_impact_curves.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+            indicator_units=item.indicator_units,
+            points_x=points_x,
+            points_y=points_y,
+            cap_of_points_x=item.cap_of_points_x,
+            cap_of_points_y=cap_of_points_y,
+            activation_of_points_x=item.activation_of_points_x,
+        )
+
+
+class PiecewiseLinearImpactCurve(ImpactCurve):
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_impact_curves.py:268]: Missing docstring in public class(fails Ruff rule D101)

File: src/physrisk/vulnerability_models/config_based_impact_curves.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+            points_x[self.cap_of_points_x < points_x] = self.cap_of_points_x
+        points_y = np.interp(points_x, self.points_x, self.points_y)
+        if self.cap_of_points_y is not None:
+            points_y[self.cap_of_points_y < points_y] = self.cap_of_points_y
+        if self.activation_of_points_x is not None:
+            points_y[points_x <= self.activation_of_points_x] = 0.0
+        return points_y
+
+
+class ParametricImpactCurve(ImpactCurve):
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_impact_curves.py:300]: Missing docstring in public class(fails Ruff rule D101)

File: src/physrisk/vulnerability_models/config_based_impact_curves.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+            curve_type,
+            indicator_units,
+            cap_of_points_x,
+            cap_of_points_y,
+            activation_of_points_x,
+        )
+        self.points_x = points_x
+        self.points_y = points_y
+
+    def get_impact(self, points_x: np.ndarray) -> np.ndarray:
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_impact_curves.py:289]: Missing docstring in public method (fails Ruff rule D102)

File: src/physrisk/vulnerability_models/config_based_impact_curves.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+        points_y = np.interp(points_x, self.points_x, self.points_y)
+        if self.cap_of_points_y is not None:
+            points_y[self.cap_of_points_y < points_y] = self.cap_of_points_y
+        if self.activation_of_points_x is not None:
+            points_y[points_x <= self.activation_of_points_x] = 0.0
+        return points_y
+
+
+class ParametricImpactCurve(ImpactCurve):
+    def __init__(
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_impact_curves.py:301]: Missing docstring in init (fails Ruff rule D107)

File: src/physrisk/vulnerability_models/config_based_impact_curves.py [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+            activation_of_points_x,
+        )
+        if self.cap_of_points_y == 0.0:
+            self.scale = 1.0
+        else:
+            self.scale = (points_x[0] - activation_of_points_x) / np.log(
+                1.0 / (1.0 - (points_y[0] / self.cap_of_points_y))
+            )
+
+    def get_impact(self, points_x: np.ndarray) -> np.ndarray:
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_impact_curves.py:329]: Missing docstring in public method (fails Ruff rule D102)

File: src/physrisk/vulnerability_models/config_based_vuln_model_acute.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
+from typing import (
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_vuln_model_acute.py:1]: Missing docstring in public module (fails Ruff rule D100).

File: src/physrisk/vulnerability_models/config_based_vuln_model_acute.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+from physrisk.vulnerability_models.config_based_impact_curves import (
+    ImpactCurve,
+    ImpactCurveKey,
+    VulnerabilityConfigItem,
+)
+
+ureg = UnitRegistry()
+
+
+class ConfigBasedVulnerabilityModelAcute(VulnerabilityModelAcuteBase):
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_vuln_model_acute.py:36]: Missing docstring in public class (fails Ruff rule D101)

File: src/physrisk/vulnerability_models/config_based_vuln_model_acute.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+    def __init__(
+        self,
+        *,
+        asset_class: str,  # type of the asset class
+        hazard_class: type,  # type of the hazard class
+        impact_id: str,
+        config_items: Sequence[VulnerabilityConfigItem],
+    ):
+        """A vulnerability model that requires only specification of a set of damage/disruption curves
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_vuln_model_acute.py:45]: 1 blank line required between summary line and description  (fails Ruff rule D205)

First line of docstring should be in imperative mood: "A vulnerability model that requires only specification of a set of damage/disruption curves" (fails Ruff rule D401)

File: src/physrisk/vulnerability_models/config_based_vuln_model_acute.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+            for item in config_items
+        }
+        self.has_sop = (hazard_class == RiverineInundation) or (
+            hazard_class == PluvialInundation
+        )
+        self.asset_attributes = set(
+            key for curve in self.curves for key in curve.identifier.keys()
+        )
+
+    def get_data_requests(
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_vuln_model_acute.py:70]: Missing docstring in public method (fails Ruff rule D102)

File: src/physrisk/vulnerability_models/config_based_vuln_model_acute.py [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+
+    def _sop_id_for_hazard(self, hazard_type: type):
+        if (hazard_type == RiverineInundation) or (hazard_type == PluvialInundation):
+            return "flood_sop"
+        else:
+            raise NotImplementedError(
+                f"sop not available for hazard {hazard_type.__name__}"
+            )
+
+    def get_distributions(
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_vuln_model_acute.py:111]: Missing docstring in public method (fails Ruff rule D102)

File: src/physrisk/vulnerability_models/config_based_vuln_model_param.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
+from typing import (
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_vuln_model_param.py:1]: Missing docstring in public module (fails Ruff rule D100).

File: src/physrisk/vulnerability_models/config_based_vuln_model_param.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+from physrisk.kernel.vulnerability_model import VulnerabilityModelBase
+from physrisk.vulnerability_models.config_based_impact_curves import (
+    ImpactCurve,
+    ImpactCurveKey,
+    VulnerabilityConfigItem,
+    to_year_fraction_divisor,
+)
+
+
+class ConfigBasedVulnerabilityModel(VulnerabilityModelBase):
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_vuln_model_param.py:28]: Missing docstring in public class (fails Ruff rule D101)

File: src/physrisk/vulnerability_models/config_based_vuln_model_param.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+    def __init__(
+        self,
+        *,
+        asset_class: str,  # type of the asset class
+        hazard_class: type,  # type of the hazard class
+        impact_id: str,
+        config_items: Sequence[VulnerabilityConfigItem],
+    ):
+        """A vulnerability model that requires only specification of a set of damage/disruption curves
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_vuln_model_param.py:37]: 1 blank line required between summary line and description (fails Ruff rule D205)
First line of docstring should be in imperative mood: "A vulnerability model that requires only specification of a set of damage/disruption curves" (fails Ruff rule D401)

File: src/physrisk/vulnerability_models/config_based_vuln_model_param.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+        )
+
+        self.thresholds = None
+        if (
+            self.hazard_type == ChronicHeat
+            and self.indicator_id == "days_tas/above/{temp_c}c"
+        ):
+            self.thresholds = np.array(range(25, 60, 5))
+
+    def get_data_requests(
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_vuln_model_param.py:70]: Missing docstring in public method (fails Ruff rule D102)

File: src/physrisk/vulnerability_models/config_based_vuln_model_param.py [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+        return HazardDataRequest(
+            self.hazard_type,
+            asset.longitude,
+            asset.latitude,
+            scenario=scenario,
+            year=year,
+            indicator_id=self.indicator_id,
+        )
+
+    def get_impact(
```
Comment (VMarfima) [inline src/physrisk/vulnerability_models/config_based_vuln_model_param.py:98]: Missing docstring in public method (fails Ruff rule D102)

File: tests/data/hazard_model_store_test.py [hunk 20]
```diff
@@ -68,9 +68,9 @@ def _crs_shape_transform_global(
         self,
         width: int = 43200,
         height: int = 21600,
-        return_periods: Union[List[float], npt.NDArray] = [0.0],
+        index_values: Union[List[float], npt.NDArray, List[str]] = [0.0],
```
Comment (VMarfima) [inline tests/data/hazard_model_store_test.py:71]: Do not use mutable data structures for argument defaults (fails Ruff rule B006).

File: tests/data/hazard_model_store_test.py [hunk 21]
```diff
@@ -118,10 +118,10 @@ def _crs_shape_transform(
         self,
         width: int,
         height: int,
-        return_periods: Union[List[float], npt.NDArray] = [0.0],
+        index_values: Union[List[float], npt.NDArray, List[str]] = [0.0],
```
Comment (VMarfima) [inline tests/data/hazard_model_store_test.py:121]: Do not use mutable data structures for argument defaults (fails Ruff rule B006).

File: docs/user-guide/vulnerability_uncertainty.ipynb [hunk 22]
```diff
@@ -0,0 +1,2184 @@
...
+   ],
+   "source": [
+    "import plotly.graph_objects as go\n",
+    "\n",
+    "fig = go.Figure()\n",
+    "for i in [0, 2, 4, 6, 8]:\n",
+    "    fig.add_scatter(x=vuln.cdf[i, :], y=vuln.impact, name=f\"Flood depth {hazard_indicator[i]}m\")\n",
+    "# also interpolate for a depth of 0.6m and display\n",
+    "interp = vuln.interpolate_cdfs(np.array([0.6]))\n",
+    "fig.add_scatter(x=interp[0, :], y=vuln.impact, name=f\"Interpolated depth 0.6m\")\n",
```
Comment (VMarfima) [inline docs/user-guide/vulnerability_uncertainty.ipynb:1120]: f-string without any placeholders (fails Ruff rule F541).

**Comment by joemoorhouse** [comment]
Hi @xbarra and @VMarfima,
Thanks for comments; as suggested I pulled out the PDM lock update into a separate PR and rebased. I took the change to add the PDM matrix CI at the same time.

A couple of changes that hit a lot of files:
1) Change of Type hint in vulnerability models from Iterable[HazardDataResponse] to Sequence[HazardDataResponse]. Probably should have been like this in the first place (not really intent to allow for generators here and length should be known), but had a case with the config-based models where it was desirable to see length of responses hence change in this PR. I also addressed some other Type hints, generally preferring Sequence over List.
Hopefully does not affect too much and straight forward to change; no good time to make such changes.
2) Starting to address some of the ruff excluded items I removed exclusion for unused imports.

Added some more doc strings, but still more to do there.

Thanks,
Joe

**Pull request merged**